import: RunSubtask can hang when generateAndSendJob is canceled

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

This is a regression introduced by PR #64767, which changed `generateAndSendJob` to call `job.ref(jobWg)` for all generated jobs before sending them to `jobToWorkerCh`.

When the context is canceled after one generated job has been sent, but before the rest of the same `jobs` slice is sent, the already-refed unsent jobs are not marked done. The caller later waits on `jobWg.Wait()` and can hang indefinitely.

The stuck stack shows `RunSubtask` blocked through `doImport`:

```text
sync.(*WaitGroup).Wait
github.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).doImport
github.com/pingcap/tidb/pkg/lightning/backend/local.(*Backend).ImportEngine
github.com/pingcap/tidb/pkg/dxf/importinto.(*writeAndIngestStepExecutor).RunSubtask
github.com/pingcap/tidb/pkg/dxf/framework/taskexecutor.(*BaseTaskExecutor).runSubtask
```

A focused unit test can reproduce it by:

1. Generating multiple region jobs in a single `generateAndSendJob` batch.
2. Letting the first job be sent to `jobToWorkerCh`.
3. Calling `job.done(jobWg)` for the first job and canceling the context.
4. Returning from `generateAndSendJob`.
5. Waiting for `jobWg.Wait()`.

Before the fix, the new unit test fails with:

```text
--- FAIL: TestGenerateAndSendJobDoneAllRefedJobsOnCancel (3.05s)
    local_test.go:3007:
        Error:       jobWg.Wait should finish after generateAndSendJob returns
        Messages:    ref'd jobs after the canceled send path must all be marked done
```

### 2. What did you expect to see? (Required)

When `generateAndSendJob` observes context cancellation after refing a batch of jobs, every refed job that was not sent to a downstream owner should be marked done before returning. Then `jobWg.Wait()` should finish and `RunSubtask` should be able to exit.

### 3. What did you see instead (Required)

Only the current unsent job is marked done on the cancellation branch. Later jobs in the same `jobs` slice were already refed by the previous loop, but they were never sent to `jobToWorkerCh` and never marked done. The leaked refs keep `jobWg.Wait()` blocked, so `RunSubtask` can remain stuck.

### 4. What is your TiDB version? (Required)

Master after PR #64767. The issue was reproduced locally on commit `c23b91914b` with the unit test above.

--- a/pkg/ingestor/ingestctrl/local.go
+++ b/pkg/ingestor/ingestctrl/local.go
@@ -1220,9 +1220,10 @@ func (local *Backend) generateAndSendJob(
 						}
 						return err
 					}
-					// we need to increase the ref count before sending jobs to
-					// jobToWorkerCh, in case some job finished quickly and decrease
-					// the ref count to zero and cause the data being released.
+					// All jobs must be ref'd before sending any job to jobToWorkerCh.
+					// Jobs run asynchronously after they are sent. If an earlier job finishes
+					// before later jobs increase their refs, the ingest data can be released
+					// when its ref count drops to zero.
 					for _, job := range jobs {
 						job.ref(jobWg)
 					}

--- a/pkg/ingestor/ingestctrl/local.go
+++ b/pkg/ingestor/ingestctrl/local.go
@@ -1226,11 +1226,13 @@ func (local *Backend) generateAndSendJob(
 					for _, job := range jobs {
 						job.ref(jobWg)
 					}
-					for _, job := range jobs {
+					for i, job := range jobs {
 						select {
 						case <-egCtx.Done():
-							// this job is not put into jobToWorkerCh
-							job.done(jobWg)
+							for _, unsentJob := range jobs[i:] {
+								// These jobs are not put into jobToWorkerCh.
+								unsentJob.done(jobWg)
+							}
 							// if the context is canceled, it means worker has error.
 							return nil
 						case jobToWorkerCh <- job:

--- a/pkg/ingestor/ingestctrl/local_test.go
+++ b/pkg/ingestor/ingestctrl/local_test.go
@@ -2937,6 +2937,78 @@ func TestRefAllJobsBeforeSending(t *testing.T) {
 	require.True(t, data.GetRefCount() == 0, "ref count should be 0 after all jobs are done")
 }
 
+func TestGenerateAndSendJobDoneAllRefedJobsOnCancel(t *testing.T) {
+	testfailpoint.Enable(t, "github.com/pingcap/tidb/pkg/ingestor/ingestctrl/fakeRegionJobs", "return()")
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+	local := &Backend{}
+	local.WorkerConcurrency.Store(1)
+
+	data := &refCountIngestData{
+		mockIngestData: mockIngestData{
+			{[]byte("b"), []byte("b")},
+			{[]byte("c"), []byte("c")},
+			{[]byte("d"), []byte("d")},
+			{[]byte("e"), []byte("e")},
+		},
+	}
+	jobs := []*regionJob{
+		{keyRange: engineapi.Range{Start: []byte("a"), End: []byte("b")}, ingestData: data, region: dummyRegionInfo},
+		{keyRange: engineapi.Range{Start: []byte("b"), End: []byte("c")}, ingestData: data, region: dummyRegionInfo},
+		{keyRange: engineapi.Range{Start: []byte("c"), End: []byte("d")}, ingestData: data, region: dummyRegionInfo},
+		{keyRange: engineapi.Range{Start: []byte("d"), End: []byte("e")}, ingestData: data, region: dummyRegionInfo},
+	}
+	jobRange := engineapi.Range{Start: []byte("a"), End: []byte("e")}
+	fakeRegionJobs = map[[2]string]struct {
+		jobs []*regionJob
+		err  error
+	}{
+		{"a", "e"}: {
+			jobs: jobs,
+		},
+	}
+	t.Cleanup(func() {
+		fakeRegionJobs = nil
+	})
+
+	mockEngine := &mockEngineWithData{
+		data:   data,
+		ranges: []engineapi.Range{jobRange},
+	}
+	jobToWorkerCh := make(chan *regionJob)
+	var jobWg sync.WaitGroup
+	firstJobDone := make(chan struct{})
+
+	go func() {
+		job := <-jobToWorkerCh
+		job.done(&jobWg)
+		cancel()
+		close(firstJobDone)
+	}()
+
+	err := local.generateAndSendJob(ctx, mockEngine, int64(config.SplitRegionSize), int64(config.SplitRegionKeys), jobToWorkerCh, &jobWg)
+	require.NoError(t, err)
+	<-firstJobDone
+
+	waitDone := make(chan struct{})
+	go func() {
+		jobWg.Wait()
+		close(waitDone)
+	}()
+
+	select {
+	case <-waitDone:
+	case <-time.After(3 * time.Second):
+		for _, job := range jobs[2:] {
+			job.done(&jobWg)
+		}
+		<-waitDone
+		require.Fail(t, "jobWg.Wait should finish after generateAndSendJob returns", "ref'd jobs after the canceled send path must all be marked done")
+	}
+	require.Equal(t, int64(0), data.GetRefCount())
+}
+
 // mockEngineWithData is a mock engine that implements engineapi.Engine
 type mockEngineWithData struct {
 	data   engineapi.IngestData

--- a/pkg/ingestor/ingestctrl/local_test.go
+++ b/pkg/ingestor/ingestctrl/local_test.go
@@ -2993,7 +2993,7 @@ func TestGenerateAndSendJobDoneAllRefedJobsOnCancel(t *testing.T) {
 
 	require.Eventually(t, func() bool {
 		return data.GetRefCount() == 0
-	}, 3*time.Second, 10*time.Millisecond, "ref'd jobs after the canceled send path must all be marked done")
+	}, 10*time.Second, 10*time.Millisecond, "ref'd jobs after the canceled send path must all be marked done")
 	jobWg.Wait()
 	require.Equal(t, int64(0), data.GetRefCount())
 }

--- a/pkg/ingestor/ingestctrl/local_test.go
+++ b/pkg/ingestor/ingestctrl/local_test.go
@@ -2991,31 +2991,10 @@ func TestGenerateAndSendJobDoneAllRefedJobsOnCancel(t *testing.T) {
 	require.NoError(t, err)
 	<-firstJobDone
 
-	waitDone := make(chan struct{})
-	go func() {
-		jobWg.Wait()
-		close(waitDone)
-	}()
-
-	cleanupLeakedJobs := true
-	defer func() {
-		if !cleanupLeakedJobs {
-			return
-		}
-		for _, job := range jobs[2:] {
-			job.done(&jobWg)
-		}
-		<-waitDone
-	}()
 	require.Eventually(t, func() bool {
-		select {
-		case <-waitDone:
-			return true
-		default:
-			return false
-		}
+		return data.GetRefCount() == 0
 	}, 3*time.Second, 10*time.Millisecond, "ref'd jobs after the canceled send path must all be marked done")
-	cleanupLeakedJobs = false
+	jobWg.Wait()
 	require.Equal(t, int64(0), data.GetRefCount())
 }
 

--- a/pkg/ingestor/ingestctrl/local_test.go
+++ b/pkg/ingestor/ingestctrl/local_test.go
@@ -2997,15 +2997,25 @@ func TestGenerateAndSendJobDoneAllRefedJobsOnCancel(t *testing.T) {
 		close(waitDone)
 	}()
 
-	select {
-	case <-waitDone:
-	case <-time.After(3 * time.Second):
+	cleanupLeakedJobs := true
+	defer func() {
+		if !cleanupLeakedJobs {
+			return
+		}
 		for _, job := range jobs[2:] {
 			job.done(&jobWg)
 		}
 		<-waitDone
-		require.Fail(t, "jobWg.Wait should finish after generateAndSendJob returns", "ref'd jobs after the canceled send path must all be marked done")
-	}
+	}()
+	require.Eventually(t, func() bool {
+		select {
+		case <-waitDone:
+			return true
+		default:
+			return false
+		}
+	}, 3*time.Second, 10*time.Millisecond, "ref'd jobs after the canceled send path must all be marked done")
+	cleanupLeakedJobs = false
 	require.Equal(t, int64(0), data.GetRefCount())
 }
 

